contracts/shared/ComponentService.sol

```diff
@@ -483,73 +493,55 @@ contract ComponentService is
     /// @dev registers the component represented by the provided address
     function _register(
         address componentAddress, // address of component to register
-        ObjectType requiredType, // required type for component for registration
-        RoleId requiredRole // role required for comonent owner for registration
+        ObjectType requiredType // required type for component for registration
     )
         internal
         virtual
         returns (
             InstanceReader instanceReader, 
+            InstanceAdmin instanceAdmin, 
             InstanceStore instanceStore, 
             NftId componentNftId
         )
     {
         (
+            NftId instanceNftId,
             IInstance instance,
             IInstanceLinkedComponent component,
-            address owner
+            address initialOwner
         ) = _getAndVerifyRegisterableComponent(
             componentAddress,
-            requiredType,
-            requiredRole);
+            requiredType);
 
-        // register component with registry
-        componentNftId = _registryService.registerComponent(
-            component, 
-            requiredType, 
-            owner).nftId;
+        // register with registry
+        if (requiredType == PRODUCT()) {
+            componentNftId = _registryService.registerProduct(
+                component, initialOwner).nftId;
+        } else {
+            componentNftId = _registryService.registerProductLinkedComponent(
+                component, requiredType, initialOwner).nftId;
+        }
 
         component.linkToRegisteredNftId();
```

Comment:
If code after linking can not be executed in a separate tx -> linking MUST be done at the most